crates/flipperzero/src/furi/string.rs

```diff
@@ -84,6 +84,13 @@ impl FuriString {
         unsafe { CStr::from_ptr(self.as_c_ptr()) }
     }
 
+    /// Raw pointer to the inner sys::FuriString
+    #[inline]
+    #[must_use]
+    pub fn as_ptr(&self) -> *mut sys::FuriString {
```

Comment:
This should take `&mut self` and be called `as_mut` since this effectively allows you mutable access to the string data.

Reply:
I would disagree. `core` and `std` types do not require `mut` on `self` since the raw pointer itself cannot be used in any bad way, while its dereferenceing is anyway `unsafe`.

Also our existing APIs use `&self`, for example `UnsafeRecord::as_ptr(&self)`

Reply:
I was a bit unsure about this method because it's just a wrapper for the `as_ptr()` method of the core `NonNull` type. `NonNull<T>::as_ptr()` takes `self` and returns an `*mut T`.

Reply:
This method absolutely needs to take `&mut self`. AFAICT the only place this PR uses this method is `sys::dialog_file_browser_show`, which mutates the provided `sys::FuriString` to store the selected path. From the Rust perspective, this is mutating the `FuriString` and therefore requires exclusive access.

I also intentionally did not add this method to the public API, because I do not want it to exist there; it enables violating the safety requirements relied on by the `FuriString` struct. We should instead be crafting safe APIs that wrap the FFI methods taking `*mut sys::FuriString`, inside which we can call this in a crate-private setting where we know the safety requirements are satisfied.

Finally, we should name this `as_mut_ptr` to match equivalent APIs in the Rust stdlib ([`slice::as_mut_ptr`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.as_mut_ptr), [`Vec::as_mut_ptr`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.as_mut_ptr), [`str::as_mut_ptr`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.as_mut_ptr)). I'd personally also prefer that this method be `unsafe`, though technically obtaining this kind of reference is safe Rust (it's just using said reference that is unsafe), so I'm not going to require that.
```suggestion
    pub(crate) fn as_mut_ptr(&mut self) -> *mut sys::FuriString {
```

Reply:
@JarvisCraft As far as I'm aware methods like [`str::as_mut_ptr`](https://doc.rust-lang.org/std/primitive.str.html#method.as_mut_ptr) all need `&mut self` (or `self` so they consume the original object).

Reply:
@str4d I think there's always going to be a need for folks to be able to get access to the underlying `*sys::FuriString` for any APIs that we haven't wrapped (e.g. maybe a third party Flipper Zero C library).

> I do not want it to exist there; it enables violating the safety requirements relied on by the `FuriString` struct

I'm trying to think of which safety requirements you could violate. My understanding was that `FuriString` was just a thin wrapper around `NonNull<sys::FuriString>`, so as long as you don't violate any standard ownership rules (e.g. calling `furi_string_free` on a string that you aren't the owner of), then it can be used safely.

I did see that our `FuriString` has the property "no nul bytes in the middle", but that might not be strictly necessary (if `m-string` allows it). While `CString` forbids an interior nul byte (I guess so the `into_raw`/`from_raw` pair works), `CStr` does not have that requirement.

Let me know if I'm missing something (which is well possible here).

Reply:
In my opinion, adding `&mut` is not the right choice here.

It does not add to the safety since you [can](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ae5ff565df8bdd967ca844eeaefb1d78) safely convert `*const T` to `*mut T`. At the same time, this requirement would add a faulty feeling in safety to its user making him think that something is guaranteed by it (which is false).

Reply:
From the other side, I like the crate-local access modifier so I generally like @str4d's suggestion.

Reply:
Actually, you can propose this change to flipperdevices/flipperzero-firmware. I've done similar job with adding `const`-qualifiers and they were fine with merging them.

Reply:
> Actually, `NonNull` is [`impl Copy`](https://doc.rust-lang.org/src/core/ptr/non_null.rs.html#701) so it does not effectively consume it.

Oh! I did not realize that was the case. OK. That makes more sense. Also I see `NonNull<T>` is defined as being the same as `*mut T`, so there's no need for a `mut self`, because `self` *is* a pointer-to-mut-T.

> In my opinion, adding `&mut` is not the right choice here.

I see this is analogous to `String::as_ptr(&self) -> *const u8` and `String::as_mut_ptr(&mut self) -> *mut u8`. Sure, you can freely cast back and forth between `*const` and `*mut` (since Rust won't let you actually _do_ anything with these pointers in safe code). However if you (the user) violate mutability rules (e.g. by creating a `mut*` to an immutable object and then mutatating it in `unsafe` code), then that is not Rust's fault. You, the user, are responsible for upholding Rust's mutability rules in `unsafe` code.

`const*` doesn't really mean all that much in Rust (or C). It's essentially type-checked documentation. About the only thing it does in Rust is [prevents you from performing mutable operations when you dereference it](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b95ce199ad8d90a13522a3c2a194dcb4).

Is it useless? No, it's still helpful to have the compiler act as a safety. The `furi_string_` methods are all const-correct and will stop you if you try to mutate something through a `const*`. If you take off the safety and point it at your foot, well you're in unsafe code, so neither Rust or C is going to stop you.

This would all be far less confusing if we could have done in-place/`#[repr(transparent)]` allocation of `FuriString` (flipperdevices/flipperzero-firmware#2461). Then it's super clear that you can't get a `*FuriString` that you can legally write through without having a `mut self` first.

> From the other side, I like the crate-local access modifier so I generally like @str4d's suggestion.

OK, I have no problem with keeping this `pub(crate)`.

Reply:
> Maybe we can include a note in the comment doc that although this method needs a mutable reference, the starting file path will not be modified. It would be better if the `sys::dialog_show_file_browser()` C function took a `*const FuriString` for path instead of `*FuriString`.

This is one totally legitimate case for `*const T as *mut T`. It's fine to call `FuriString::as_ptr() -> *const sys::FuriString` and then cast it to a `*mut sys::FuriString` so long as you don't modify memory through this pointer.

Ultimately what matters (for correct compilation) is "does `sys::dialog_show_file_browser()` actually mutate the data behind the `path` pointer?", not whether the function declaration says `const *FuriString` or `*FuriString`.